// serializable-types/dist/typeComposers/array.js
"use strict";

var decorateTypeDef = require("../decorateTypeDef");

module.exports = function array(memberDef) {
  return decorateTypeDef({
    description: "Array<".concat(memberDef.description, ">"),
    serializedDescription: "{ $type: \"array\", $value: Array<".concat(memberDef.serializedDescription, "> }"),
    check: function check(val) {
      return Array.isArray(val) && val.every(function (member) {
        return memberDef.check(member);
      });
    },
    serialize: function serialize(array) {
      return {
        $type: "array",
        $value: array.map(function (member) {
          return memberDef.serialize(member);
        })
      };
    },
    checkSerialized: function checkSerialized(serialized) {
      return serialized.$type === "array" && serialized.$value.every(function (serializedMember) {
        return memberDef.checkSerialized(serializedMember);
      });
    },
    deserialize: function deserialize(serialized) {
      return serialized.$value.map(function (serializedValue) {
        return memberDef.deserialize(serializedValue);
      });
    }
  });
};
