// serializable-types/dist/typeComposers/set.js
"use strict";

var decorateTypeDef = require("../decorateTypeDef");

module.exports = function set(memberDef) {
  return decorateTypeDef({
    description: "Set<".concat(memberDef.description, ">"),
    serializedDescription: "{ $type: \"Set\", $value: Array<".concat(memberDef.serializedDescription, "> }"),
    check: function check(val) {
      return val instanceof Set && Array.from(val).every(function (member) {
        return memberDef.check(member);
      });
    },
    serialize: function serialize(set) {
      return {
        $type: "Set",
        $value: Array.from(set).map(function (member) {
          return memberDef.serialize(member);
        })
      };
    },
    checkSerialized: function checkSerialized(serialized) {
      return serialized.$type === "Set" && serialized.$value.every(function (serializedMember) {
        return memberDef.checkSerialized(serializedMember);
      });
    },
    deserialize: function deserialize(serialized) {
      return new Set(serialized.$value.map(function (serializedValue) {
        return memberDef.deserialize(serializedValue);
      }));
    }
  });
};
